File: engine/dynamic/render_vulkan/src/util/texture.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/logging.hpp"

#include "argus/resman/resource.hpp"

#include "argus/render/common/material.hpp"

#include "internal/render_vulkan/setup/device.hpp"
#include "internal/render_vulkan/state/renderer_state.hpp"
#include "internal/render_vulkan/util/buffer.hpp"
#include "internal/render_vulkan/util/command_buffer.hpp"
#include "internal/render_vulkan/util/image.hpp"
#include "internal/render_vulkan/util/texture.hpp"

namespace argus {
    PreparedTexture prepare_texture(const LogicalDevice &device, const CommandBufferInfo &cmd_buf,
            const Resource &texture_res) {
        const auto &texture = texture_res.get<TextureData>();

        uint32_t channels = 4;
        VkDeviceSize image_size = texture.width * texture.height * channels;

        auto format = VK_FORMAT_R8G8B8A8_SRGB;

        auto image = create_image_and_image_view(device, format, { texture.width, texture.height },
                VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_SAMPLED_BIT, VK_IMAGE_ASPECT_COLOR_BIT);

        auto staging_buf = alloc_buffer(device, image_size, VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
                VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);

        {
            auto buf_mapped = map_buffer(device, staging_buf, 0, image_size, 0);
            const size_t bytes_per_pixel = 4;
            const size_t bytes_per_row = texture.width * bytes_per_pixel;
            for (size_t y = 0; y < texture.height; y++) {
                auto dst = reinterpret_cast<void *>(reinterpret_cast<uintptr_t>(buf_mapped) + (y * bytes_per_row));
                memcpy(dst, texture.get_pixel_data()[y], bytes_per_row);
            }
            unmap_buffer(device, staging_buf);
        }

        VkBufferImageCopy region{};
        region.bufferOffset = 0;
        region.bufferRowLength = 0;
        region.bufferImageHeight = 0;

        region.imageSubresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
        region.imageSubresource.mipLevel = 0;
        region.imageSubresource.baseArrayLayer = 0;
        region.imageSubresource.layerCount = 1;

        region.imageOffset = {0, 0, 0};
        region.imageExtent = { texture.width, texture.height, 1 };

        perform_image_transition(cmd_buf, image,
                VK_IMAGE_LAYOUT_UNDEFINED, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
                0, VK_ACCESS_TRANSFER_WRITE_BIT,
                VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT);

        vkCmdCopyBufferToImage(cmd_buf.handle, staging_buf.handle, image.handle,
                VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &region);

        perform_image_transition(cmd_buf, image,
                VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL,
                VK_ACCESS_TRANSFER_WRITE_BIT, VK_ACCESS_SHADER_READ_BIT,
                VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT);

        VkSamplerCreateInfo sampler_info{};
        sampler_info.sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO;
        sampler_info.magFilter = VK_FILTER_NEAREST;
        sampler_info.minFilter = VK_FILTER_NEAREST;
        sampler_info.addressModeU = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE;
        sampler_info.addressModeV = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE;
        sampler_info.addressModeW = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE;
        sampler_info.anisotropyEnable = VK_FALSE;
        sampler_info.maxAnisotropy = 0;
        sampler_info.borderColor = VK_BORDER_COLOR_INT_OPAQUE_BLACK;
        sampler_info.unnormalizedCoordinates = VK_FALSE;
        sampler_info.compareEnable = VK_FALSE;
        sampler_info.compareOp = VK_COMPARE_OP_ALWAYS;
        sampler_info.mipmapMode = VK_SAMPLER_MIPMAP_MODE_LINEAR;
        sampler_info.mipLodBias = 0.0f;
        sampler_info.minLod = 0.0f;
        sampler_info.maxLod = 0.0f;

        VkSampler sampler;
        vkCreateSampler(device.logical_device, &sampler_info, nullptr, &sampler);

        return { texture_res.uid, image, sampler, staging_buf };
    }

    void get_or_load_texture(RendererState &state, const Resource &material_res) {
        auto &texture_uid = material_res.get<Material>().get_texture_uid();

        auto existing_it = state.prepared_textures.find(texture_uid);
        if (existing_it != state.prepared_textures.end()) {
            existing_it->second.acquire();
            state.material_textures.insert({material_res.uid, texture_uid});
            return;
        }

        auto &texture_res = ResourceManager::instance().get_resource(texture_uid);

        auto prepared = prepare_texture(state.device, state.copy_cmd_buf, texture_res);

        texture_res.release();

        state.prepared_textures.insert({ texture_uid, prepared });
        state.material_textures.insert({ material_res.uid, texture_uid });

        state.texture_bufs_to_free.push_back(prepared.staging_buf);
    }

    void destroy_texture(const LogicalDevice &device, const PreparedTexture &texture) {
        vkDestroySampler(device.logical_device, texture.sampler, nullptr);
        destroy_image_and_image_view(device, texture.image);
    }
}

File: engine/modules/render_vulkan/src/util/texture.rs
use crate::setup::device::VulkanDevice;
use crate::setup::instance::VulkanInstance;
use crate::util::{CommandBufferInfo, VulkanBuffer, VulkanImage, MEM_CLASS_DEVICE_RW};
use argus_render::common::TextureData;
use argus_resman::Resource;
use argus_util::math::Vector2u;
use ash::vk;

pub(crate) struct PreparedTexture {
    uid: String,
    pub(crate) image: VulkanImage,
    pub(crate) sampler: vk::Sampler,
    staging_buf: VulkanBuffer,
}

pub(crate) fn prepare_texture(
    instance: &VulkanInstance,
    device: &VulkanDevice,
    cmd_buf: &CommandBufferInfo,
    texture_res: Resource,
) -> Result<PreparedTexture, String> {
    let texture = texture_res.get::<TextureData>().unwrap();

    let channels = 4;
    let image_size = (texture.get_width() * texture.get_height() * channels) as vk::DeviceSize;

    let format = vk::Format::R8G8B8A8_SRGB;

    let image = VulkanImage::create_image_with_view(
        instance,
        device,
        format,
        Vector2u::new(texture.get_width(), texture.get_height()),
        vk::ImageUsageFlags::TRANSFER_DST | vk::ImageUsageFlags::SAMPLED,
        vk::ImageAspectFlags::COLOR
    )?;

    let mut staging_buf = VulkanBuffer::new(
        instance,
        device,
        image_size,
        vk::BufferUsageFlags::TRANSFER_SRC,
        MEM_CLASS_DEVICE_RW,
    )?;

    {
        let bytes_per_pixel = 4;
        let bytes_per_row = texture.get_width() * bytes_per_pixel;
        let mut mapped = staging_buf.map(device, 0, vk::WHOLE_SIZE, vk::MemoryMapFlags::empty())
            .map_err(|err| err.to_string())?;
        for y in 0..texture.get_height() {
            let dst = mapped.offset_mut(y * bytes_per_row);
            dst.copy_from_slice(&texture.get_pixel_data()[(y * bytes_per_row) as usize..]);
        }
    }

    let image_sr_layers = vk::ImageSubresourceLayers::default()
        .aspect_mask(vk::ImageAspectFlags::COLOR)
        .mip_level(0)
        .base_array_layer(0)
        .layer_count(1);
    let region = vk::BufferImageCopy::default()
        .buffer_offset(0)
        .buffer_row_length(0)
        .buffer_image_height(0)
        .image_subresource(image_sr_layers)
        .image_offset(vk::Offset3D { x: 0, y: 0, z: 0 })
        .image_extent(vk::Extent3D {
            width: texture.get_width(),
            height: texture.get_height(),
            depth: 1,
        });

    image.perform_transition(
        device,
        cmd_buf,
        vk::ImageLayout::UNDEFINED,
        vk::ImageLayout::TRANSFER_DST_OPTIMAL,
        vk::AccessFlags::empty(),
        vk::AccessFlags::TRANSFER_WRITE,
        vk::PipelineStageFlags::TOP_OF_PIPE,
        vk::PipelineStageFlags::TRANSFER,
    );

    unsafe {
        device.logical_device.cmd_copy_buffer_to_image(
            cmd_buf.get_handle(),
            staging_buf.get_handle(),
            image.get_vk_image(),
            vk::ImageLayout::TRANSFER_DST_OPTIMAL,
            &[region],
        );
    }

    image.perform_transition(
        device,
        cmd_buf,
        vk::ImageLayout::TRANSFER_DST_OPTIMAL,
        vk::ImageLayout::SHADER_READ_ONLY_OPTIMAL,
        vk::AccessFlags::TRANSFER_WRITE,
        vk::AccessFlags::SHADER_READ,
        vk::PipelineStageFlags::TRANSFER,
        vk::PipelineStageFlags::FRAGMENT_SHADER,
    );

    let sampler_info = vk::SamplerCreateInfo::default()
        .mag_filter(vk::Filter::NEAREST)
        .min_filter(vk::Filter::NEAREST)
        .address_mode_u(vk::SamplerAddressMode::CLAMP_TO_EDGE)
        .address_mode_v(vk::SamplerAddressMode::CLAMP_TO_EDGE)
        .address_mode_w(vk::SamplerAddressMode::CLAMP_TO_EDGE)
        .anisotropy_enable(false)
        .max_anisotropy(0.0)
        .border_color(vk::BorderColor::INT_OPAQUE_BLACK)
        .unnormalized_coordinates(false)
        .compare_enable(false)
        .compare_op(vk::CompareOp::ALWAYS)
        .mipmap_mode(vk::SamplerMipmapMode::LINEAR)
        .mip_lod_bias(0.0)
        .min_lod(0.0)
        .max_lod(0.0);

    let sampler = unsafe {
        device.logical_device.create_sampler(&sampler_info, None)
            .map_err(|err| err.to_string())?
    };

    Ok(PreparedTexture {
        uid: texture_res.get_prototype().uid.to_string(),
        image,
        sampler,
        staging_buf,
    })
}

pub(crate) fn destroy_texture(device: &VulkanDevice, texture: PreparedTexture) {
    unsafe { device.logical_device.destroy_sampler(texture.sampler, None) };
    texture.image.destroy(device);
    texture.staging_buf.destroy(device);
}
